### PR TITLE
implement EnqueueExtensions interface in taint toleration scheduling

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind feature
/sig scheduling

#### What this PR does / why we need it:
Implement EnqueueExtensions interface in  taint toleration plugin.

#### Which issue(s) this PR fixes:

Part of #94009

#### Special notes for your reviewer:
**Q: Why this plugin should be interested in the specified events only?**
**A:** If a Pod failed by taint/toleration plugin, add/deleted or update event of an assigned Pod wouldn't help as that would only move Pods failed by podAffinity related plugins. But the node taint update would make node availble for the Pod. Also, adding a new Node, or capacity change of an existing Node may help to make the pod schedulable.

**Q: Any performance test proving this?**
**A:** Locally by leveraging #98900, I designed a testcase like this:
```
- name: SchedulingWithTolerationTaint
  workloadTemplate:
  - opcode: createNodes
    countParam: $initNodes
    templatePaths:
    - config/churn/node-with-taint.yaml
  - opcode: createPods
    countParam: $initPods
    podTemplatePath: config/pod-high-priority.yaml // without node affinity
    skipWaitToCompletion: true
  - opcode: churn
    mode: recreate
    number: 1
    templatePaths:
    - config/churn/service-default.yaml
    intervalMilliseconds: 1000
  - opcode: createPods
    countParam: $measurePods
    podTemplatePath: config/pod-with-tolerations.yaml
    collectMetrics: true
  workloads:
  - name: 1000Nodes
    params:
      initNodes: 1000
      initPods: 200
      measurePods: 1000
  - name: 5000Nodes
    params:
      initNodes: 5000
      initPods: 1000
      measurePods: 2000
Merge state
```

```
➜  k8s-sched-perf-stat git:(main) ✗ go run *.go /tmp/old.txt /tmp/new.txt
+--------------------------+-----------------------------------------+----------+--------+---------+---------+---------+
|          METRIC          |                  GROUP                  | QUANTILE |  UNIT  |   OLD   |   NEW   |  DIFF   |
+--------------------------+-----------------------------------------+----------+--------+---------+---------+---------+
| SchedulingThroughput     | SchedulingWithTolerationTaint/1000Nodes | Average  | pods/s |  153.58 |  170.40 | +10.95% |
| SchedulingThroughput     | SchedulingWithTolerationTaint/1000Nodes | Perc50   | pods/s |  165.56 |  168.11 | +1.54%  |
| SchedulingThroughput     | SchedulingWithTolerationTaint/1000Nodes | Perc90   | pods/s |  265.89 |  307.11 | +15.50% |
| SchedulingThroughput     | SchedulingWithTolerationTaint/1000Nodes | Perc99   | pods/s |  265.89 |  307.11 | +15.50% |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/1000Nodes | Average  | ms     |  322.83 |  369.86 | +14.57% |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc50   | ms     |   23.04 |   22.47 | -2.47%  |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc90   | ms     | 1271.11 | 1362.43 | +7.18%  |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc99   | ms     | 1746.20 | 1809.32 | +3.61%  |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/1000Nodes | Average  | ms     |  997.76 |  894.71 | -10.33% |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc50   | ms     |  834.44 |  735.28 | -11.88% |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc90   | ms     | 2276.28 | 2105.07 | -7.52%  |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc99   | ms     | 2843.59 | 2788.97 | -1.92%  |
| SchedulingThroughput     | SchedulingWithTolerationTaint/5000Nodes | Average  | pods/s |  168.87 |  183.73 | +8.80%  |
| SchedulingThroughput     | SchedulingWithTolerationTaint/5000Nodes | Perc50   | pods/s |  183.56 |  193.20 | +5.25%  |
| SchedulingThroughput     | SchedulingWithTolerationTaint/5000Nodes | Perc90   | pods/s |  243.11 |  259.70 | +6.82%  |
| SchedulingThroughput     | SchedulingWithTolerationTaint/5000Nodes | Perc99   | pods/s |  251.62 |  272.00 | +8.10%  |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/5000Nodes | Average  | ms     |  352.06 |  323.69 | -8.06%  |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc50   | ms     |   24.48 |   23.81 | -2.74%  |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc90   | ms     | 1510.32 | 1471.32 | -2.58%  |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc99   | ms     | 2205.06 | 1909.45 | -13.41% |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Average  | ms     | 2032.59 | 2037.57 | +0.24%  |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc50   | ms     | 2054.50 | 1946.82 | -5.24%  |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc90   | ms     | 4361.02 | 4515.75 | +3.55%  |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc99   | ms     | 4934.51 | 5338.13 | +8.18%  |
+--------------------------+-----------------------------------------+----------+--------+---------+---------+---------+
➜  k8s-sched-perf-stat git:(main) ✗ go run *.go /tmp/old.txt
+--------------------------+-----------------------------------------+----------+--------+---------+
|          METRIC          |                  GROUP                  | QUANTILE |  UNIT  | MEDIAN  |
+--------------------------+-----------------------------------------+----------+--------+---------+
| SchedulingThroughput     | SchedulingWithTolerationTaint/1000Nodes | Average  | pods/s |  153.58 |
| SchedulingThroughput     | SchedulingWithTolerationTaint/1000Nodes | Perc50   | pods/s |  165.56 |
| SchedulingThroughput     | SchedulingWithTolerationTaint/1000Nodes | Perc90   | pods/s |  265.89 |
| SchedulingThroughput     | SchedulingWithTolerationTaint/1000Nodes | Perc99   | pods/s |  265.89 |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/1000Nodes | Average  | ms     |  322.83 |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc50   | ms     |   23.04 |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc90   | ms     | 1271.11 |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc99   | ms     | 1746.20 |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/1000Nodes | Average  | ms     |  997.76 |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc50   | ms     |  834.44 |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc90   | ms     | 2276.28 |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc99   | ms     | 2843.59 |
| SchedulingThroughput     | SchedulingWithTolerationTaint/5000Nodes | Average  | pods/s |  168.87 |
| SchedulingThroughput     | SchedulingWithTolerationTaint/5000Nodes | Perc50   | pods/s |  183.56 |
| SchedulingThroughput     | SchedulingWithTolerationTaint/5000Nodes | Perc90   | pods/s |  243.11 |
| SchedulingThroughput     | SchedulingWithTolerationTaint/5000Nodes | Perc99   | pods/s |  251.62 |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/5000Nodes | Average  | ms     |  352.06 |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc50   | ms     |   24.48 |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc90   | ms     | 1510.32 |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc99   | ms     | 2205.06 |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Average  | ms     | 2032.59 |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc50   | ms     | 2054.50 |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc90   | ms     | 4361.02 |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc99   | ms     | 4934.51 |
+--------------------------+-----------------------------------------+----------+--------+---------+
```

Further yaml samples.
```
cat test/integration/scheduler_perf/config/pod-with-tolerations.yaml
apiVersion: v1
kind: Pod
metadata:
  generateName: node-toleration-
spec:
  containers:
  - image: k8s.gcr.io/pause:3.4.1
    name: pause
    ports:
    - containerPort: 80
    resources:
      limits:
        cpu: 100m
        memory: 500Mi
      requests:
        cpu: 100m
        memory: 500Mi
  tolerations:
  - key: "key1"
    operator: "Equal"
    value: "value1"
    effect: "NoExecute"
    tolerationSeconds: 3600

cat test/integration/scheduler_perf/config/node-with-taint.yaml
apiVersion: v1
kind: Node
metadata:
  generateName: scheduler-perf-
spec:
  taints:
  - key: "foo1"
    value: "value1"
    effect: "NoExecute"
status:
  capacity:
    pods: "110"
    cpu: "4"
    memory: 32Gi
  conditions:
    - status: "True"
      type: Ready
  phase: Running
```

WIP
- test case 
- perf diff
- scheduler_e2e_scheduling
- bench mark

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go [hunk 1]
```diff
@@ -50,6 +50,14 @@ func (pl *TaintToleration) Name() string {
 	return Name
 }
 
+// EventsToRegister returns the possible events that can make a Pod
+// failed by current plugin schedulable.
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go:54]:
```suggestion
// EventsToRegister returns the possible events that may make a Pod
// failed by this plugin schedulable.
```

File: pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go [hunk 2]
```diff
@@ -50,6 +50,14 @@ func (pl *TaintToleration) Name() string {
...
+func (f *Fit) EventsToRegister() []framework.ClusterEvent {
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go:55]: It's not `Fit` :)

```suggestion
func (f *TaintToleration) EventsToRegister() []framework.ClusterEvent {
```

**Review by Huang-Wei — COMMENTED** [review]
Let's add a verification:

```
var _ framework.EnqueueExtensions = &TaintToleration{}
```

**Comment by Huang-Wei** [comment]
@pacoxu to show the performance boost, you can design a testcase like this:

- Init N high-priority Pods with node-affinity.
- Init 1000/5000 Nodes, all tainted `foo=bar`.
- Use the churn operator to spawn irrelevant events regularly (like service, or pv, or pvc events)
- Measure M regular Pods with toleration `foo=bar`.

It's expected to see better perf b/c the N high-priority Pods won't frequently be moved from unscheudlableQ to activeQ upon the irrelevant events.

**Comment by Huang-Wei** [comment]
@pacoxu Some comments:

> 1. I may create 1000 node with taint foo1.
Yes

> 2. Then create 1000 pod with node affinity to the node label, but with toleration to foo2.
It's ok to remove node affinity, with toleration to foo2 or without toleration is either fine.

> 3. Create/Update Pod (which may not affect)
> 4. Update node taint to foo2 or remove those taints
> 5. Add 1000 Nodes without taints?

I think we can simplify a bit: create Services every 1 (or 2) seconds, which will churn the high-priority Pods; also create N measurePods which have tolerations matched the created Nodes.

So the yaml can look like this:

```yaml
- name: SchedulingWithTolerationTaint
  workloadTemplate:
  - opcode: createNodes
    countParam: $initNodes
    templatePaths:
    - config/churn/node-with-taint.yaml
  - opcode: createPods
    countParam: $initPods
    podTemplatePath: config/pod-high-priority.yaml // without node affinity
    skipWaitToCompletion: true
  - opcode: churn
    mode: recreate
    number: 1
    templatePaths:
    - config/churn/service-default.yaml
    intervalMilliseconds: 1000
  - opcode: createPods
    countParam: $measurePods
    podTemplatePath: config/pod-with-tolerations.yaml
    collectMetrics: true
  workloads:
  - name: 1000Nodes
    params:
      initNodes: 1000
      initPods: 200
      measurePods: 1000
  - name: 5000Nodes
    params:
      initNodes: 5000
      initPods: 1000
      measurePods: 2000
```

**Comment by pacoxu** [comment]
@Huang-Wei  I updated the test result in description

**Comment by Huang-Wei** [comment]
Thanks @pacoxu . The result looks good, but not that good as I expected... esp.:

```
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Average  | ms     | 2032.59 | 2037.57 | +0.24%  |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc50   | ms     | 2054.50 | 1946.82 | -5.24%  |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc90   | ms     | 4361.02 | 4515.75 | +3.55%  |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc99   | ms     | 4934.51 | 5338.13 | +8.18%  |
+--------------------------+-----------------------------------------+----------+--------+---------+---------+---------+
```

I will run the test later.

**Comment by pacoxu** [comment]
I run in my MaC, and I'm not stopping working on other things meanwhile.  😄

**Comment by Huang-Wei** [comment]
> I run in my MaC, and I'm not stopping working on other things meanwhile.

Oops, that can be the reason... I also used Mac to run the tests by running the tests delicately, and also turn off irrelevant apps and notifications. Nvm, let me run the test tomorrow.

**Comment by Huang-Wei** [comment]
I got a chance to run teh test today. @pacoxu Here are some changes I did based on your testcase:

- compose the taint/toleration with effect NoSchedule (instead of NoExecut)
- use `create` (instead of `recreate`) to continuously create Services in the churn operator
- run 1000 tests 20 times, and 5000 tests 5 times

Here is the result:

```
+--------------------------+-----------------------------------------+----------+--------+---------+---------+---------+
|          METRIC          |                  GROUP                  | QUANTILE |  UNIT  |   OLD   |   NEW   |  DIFF   |
+--------------------------+-----------------------------------------+----------+--------+---------+---------+---------+
| SchedulingThroughput     | SchedulingWithTolerationTaint/1000Nodes | Average  | pods/s |  227.11 |  214.57 | -5.52%  |
| SchedulingThroughput     | SchedulingWithTolerationTaint/1000Nodes | Perc50   | pods/s |  149.83 |  171.12 | +14.20% |
| SchedulingThroughput     | SchedulingWithTolerationTaint/1000Nodes | Perc90   | pods/s |  451.05 |  420.52 | -6.77%  |
| SchedulingThroughput     | SchedulingWithTolerationTaint/1000Nodes | Perc99   | pods/s |  451.05 |  420.52 | -6.77%  |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/1000Nodes | Average  | ms     |  411.95 |  446.71 | +8.44%  |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc50   | ms     |   14.11 |   14.30 | +1.33%  |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc90   | ms     | 1121.50 | 1199.86 | +6.99%  |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc99   | ms     | 1606.47 | 1694.20 | +5.46%  |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/1000Nodes | Average  | ms     |  643.50 |  651.84 | +1.30%  |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc50   | ms     |  366.96 |  338.61 | -7.72%  |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc90   | ms     | 1763.92 | 1766.06 | +0.12%  |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/1000Nodes | Perc99   | ms     | 2305.27 | 2304.87 | -0.02%  |
+--------------------------+-----------------------------------------+----------+--------+---------+---------+---------+
```

```
+--------------------------+-----------------------------------------+----------+--------+---------+---------+---------+
|          METRIC          |                  GROUP                  | QUANTILE |  UNIT  |   OLD   |   NEW   |  DIFF   |
+--------------------------+-----------------------------------------+----------+--------+---------+---------+---------+
| SchedulingThroughput     | SchedulingWithTolerationTaint/5000Nodes | Average  | pods/s |  214.08 |  277.75 | +29.74% |
| SchedulingThroughput     | SchedulingWithTolerationTaint/5000Nodes | Perc50   | pods/s |  159.20 |  258.20 | +62.19% |
| SchedulingThroughput     | SchedulingWithTolerationTaint/5000Nodes | Perc90   | pods/s |  452.60 |  503.60 | +11.27% |
| SchedulingThroughput     | SchedulingWithTolerationTaint/5000Nodes | Perc99   | pods/s |  472.20 |  503.60 | +6.65%  |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/5000Nodes | Average  | ms     |  582.41 |  425.87 | -26.88% |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc50   | ms     |   17.31 |   13.64 | -21.23% |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc90   | ms     | 1663.49 | 1155.98 | -30.51% |
| scheduler_e2e_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc99   | ms     | 2701.76 | 1908.93 | -29.34% |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Average  | ms     | 1130.19 |  786.85 | -30.38% |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc50   | ms     |  790.77 |  489.33 | -38.12% |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc90   | ms     | 2819.55 | 1977.76 | -29.86% |
| scheduler_pod_scheduling | SchedulingWithTolerationTaint/5000Nodes | Perc99   | ms     | 3881.98 | 2694.49 | -30.59% |
+--------------------------+-----------------------------------------+----------+--------+---------+---------+---------+
```

**Comment by pacoxu** [comment]
Need I run it again? @Huang-Wei

**Comment by Huang-Wei** [comment]
The 5000Nodes test looks promising.

/triage accepted
/priority important-longterm
/lgtm
/approve

**Comment by Huang-Wei** [comment]
> Need I run it again? @Huang-Wei

Not quite. The result is good enough.

**Pull request merged**